// oceanic.js/dist/lib/structures/Application.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/** @module Application */
const ClientApplication_1 = tslib_1.__importDefault(require("./ClientApplication"));
const OAuthGuild_1 = tslib_1.__importDefault(require("./OAuthGuild"));
const Routes = tslib_1.__importStar(require("../util/Routes"));
/** Represents an application. */
class Application extends ClientApplication_1.default {
    /** The approximate number of guilds the application is in. */
    approximateGuildCount;
    /** This application's rich presence invite cover image hash, if any. */
    coverImage;
    /** This application's default custom authorization link, if any. */
    customInstallURL;
    /** The description of the application. */
    description;
    /** If this application is a game sold on Discord, the guild to which it has been linked. This will only be present if recieved via `/applications/@me` {@link REST/Miscellaneous.getApplication | Miscellaneous#getApplication}). */
    guild;
    /** If this application is a game sold on Discord, the ID of the guild to which it has been linked. */
    guildID;
    /** The icon hash of the application. */
    icon;
    /** Settings for this application's in-app authorization link, if enabled. */
    installParams;
    /** This applications interaction endpoint url, if any. */
    interactionsEndpointURL;
    /** The name of the application. */
    name;
    /** If this application is a game sold on Discord, the id of the Game's SKU. */
    primarySKUID;
    /** A URL to this application's privacy policy. */
    privacyPolicyURL;
    /** This application's role connections verification url, if any. */
    roleConnectionsVerificationURL;
    /** A list of rpc origin urls, if rpc is enabled. */
    rpcOrigins;
    /** If this application is a game sold on Discord, the slug that links to its store page. */
    slug;
    /** The tags for this application. */
    tags;
    /** A URL to this application's terms of service. */
    termsOfServiceURL;
    /** The type of this application. */
    type;
    /** The bot's hex encoded public key. */
    verifyKey;
    constructor(data, client) {
        super(data, client);
        this.approximateGuildCount = data.approximate_guild_count ?? 0;
        this.coverImage = null;
        this.description = data.description;
        this.guild = null;
        this.guildID = data.guild_id ?? null;
        this.icon = null;
        this.interactionsEndpointURL = null;
        this.name = data.name;
        this.roleConnectionsVerificationURL = null;
        this.rpcOrigins = [];
        this.type = null;
        this.verifyKey = data.verify_key;
        this.update(data);
    }
    update(data) {
        super.update(data);
        if (data.approximate_guild_count !== undefined) {
            this.approximateGuildCount = data.approximate_guild_count;
        }
        if (data.cover_image !== undefined) {
            this.coverImage = data.cover_image;
        }
        if (data.custom_install_url !== undefined) {
            this.customInstallURL = data.custom_install_url;
        }
        if (data.description !== undefined) {
            this.description = data.description;
        }
        if (data.guild !== undefined) {
            this.guild = new OAuthGuild_1.default(data.guild, this.client);
        }
        this.guildID = data.guild_id === undefined ? null : data.guild_id;
        if (data.icon !== undefined) {
            this.icon = data.icon;
        }
        if (data.install_params !== undefined) {
            this.installParams = data.install_params;
        }
        if (data.interactions_endpoint_url !== undefined) {
            this.interactionsEndpointURL = data.interactions_endpoint_url;
        }
        if (data.name !== undefined) {
            this.name = data.name;
        }
        if (data.primary_sku_id !== undefined) {
            this.primarySKUID = data.primary_sku_id;
        }
        if (data.privacy_policy_url !== undefined) {
            this.privacyPolicyURL = data.privacy_policy_url;
        }
        if (data.role_connections_verification_url !== undefined) {
            this.roleConnectionsVerificationURL = data.role_connections_verification_url;
        }
        if (data.rpc_origins !== undefined) {
            this.rpcOrigins = data.rpc_origins;
        }
        if (data.slug !== undefined) {
            this.slug = data.slug;
        }
        if (data.tags !== undefined) {
            this.tags = data.tags;
        }
        if (data.terms_of_service_url !== undefined) {
            this.termsOfServiceURL = data.terms_of_service_url;
        }
        if (data.type !== undefined) {
            this.type = data.type;
        }
        if (data.verify_key !== undefined) {
            this.verifyKey = data.verify_key;
        }
    }
    /**
     * The url of this application's cover image.
     * @param format The format the url should be.
     * @param size The dimensions of the image.
     */
    coverImageURL(format, size) {
        return this.coverImage === null ? null : this.client.util.formatImage(Routes.APPLICATION_COVER(this.id, this.coverImage), format, size);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            approximateGuildCount: this.approximateGuildCount,
            coverImage: this.coverImage,
            customInstallURL: this.customInstallURL,
            description: this.description,
            guild: this.guild?.toJSON() ?? null,
            guildID: this.guildID,
            icon: this.icon,
            installParams: this.installParams,
            interactionsEndpointURL: this.interactionsEndpointURL,
            name: this.name,
            primarySKUID: this.primarySKUID,
            privacyPolicyURL: this.privacyPolicyURL,
            roleConnectionsVerificationURL: this.roleConnectionsVerificationURL,
            rpcOrigins: this.rpcOrigins,
            slug: this.slug,
            tags: this.tags,
            termsOfServiceURL: this.termsOfServiceURL,
            type: this.type,
            verifyKey: this.verifyKey
        };
    }
}
exports.default = Application;
