// oceanic.js/dist/lib/structures/InviteGuild.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/** @module InviteGuild */
const Base_1 = tslib_1.__importDefault(require("./Base"));
const Routes = tslib_1.__importStar(require("../util/Routes"));
/** Represents a guild received via an invite. */
class InviteGuild extends Base_1.default {
    _cachedCompleteGuild;
    /** The hash of this guild's banner. */
    banner;
    /** The description of this guild. */
    description;
    /** The [features](https://discord.com/developers/docs/resources/guild#guild-object-guild-features) this guild has. */
    features;
    /** The icon hash of this guild. */
    icon;
    /** The name of this guild. */
    name;
    /** The [nsfw level](https://discord.com/developers/docs/resources/guild#guild-object-guild-nsfw-level) of this guild. */
    nsfwLevel;
    /** The number of nitro boosts this guild has. */
    premiumSubscriptionCount;
    /** The invite splash hash of this guild. */
    splash;
    /** The vanity url of this guild. Only present in guilds with the `VANITY_URL` feature. */
    vanityURLCode;
    /** The [verification level](https://discord.com/developers/docs/resources/guild#guild-object-verification-level) of this guild. */
    verificationLevel;
    constructor(data, client) {
        super(data.id, client);
        this.banner = data.banner;
        this.description = data.description;
        this.features = data.features;
        this.name = data.name;
        this.nsfwLevel = data.nsfw_level;
        this.icon = data.icon;
        this.premiumSubscriptionCount = data.premium_subscription_count;
        this.splash = data.splash;
        this.vanityURLCode = data.vanity_url_code;
        this.verificationLevel = data.verification_level;
    }
    /** The complete guild this InviteGuild represents, if cached. */
    get completeGuild() {
        return this._cachedCompleteGuild ??= this.client.guilds.get(this.id);
    }
    /**
     * The url of this guild's banner.
     * @param format The format the url should be.
     * @param size The dimensions of the image.
     */
    bannerURL(format, size) {
        return this.banner === null ? null : this.client.util.formatImage(Routes.BANNER(this.id, this.banner), format, size);
    }
    /**
     * The url of this guild's icon.
     * @param format The format the url should be.
     * @param size The dimensions of the image.
     */
    iconURL(format, size) {
        return this.icon === null ? null : this.client.util.formatImage(Routes.GUILD_ICON(this.id, this.icon), format, size);
    }
    /**
     * The url of this guild's invite splash.
     * @param format The format the url should be.
     * @param size The dimensions of the image.
     */
    splashURL(format, size) {
        return this.splash === null ? null : this.client.util.formatImage(Routes.GUILD_SPLASH(this.id, this.splash), format, size);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            banner: this.banner,
            description: this.description,
            features: this.features,
            icon: this.icon,
            name: this.name,
            nsfwLevel: this.nsfwLevel,
            premiumSubscriptionCount: this.premiumSubscriptionCount,
            splash: this.splash,
            vanityURLCode: this.vanityURLCode,
            verificationLevel: this.verificationLevel
        };
    }
}
exports.default = InviteGuild;
