// oceanic.js/dist/lib/structures/StageChannel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const TextableVoiceChannel_1 = tslib_1.__importDefault(require("./TextableVoiceChannel"));
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
/** Represents a guild stage channel. */
class StageChannel extends TextableVoiceChannel_1.default {
    constructor(data, client) {
        super(data, client);
    }
    /**
     * Create a stage instance on this channel.
     * @param options The options for creating the stage instance.
     */
    async createStageInstance(options) {
        return this.client.rest.channels.createStageInstance(this.id, options);
    }
    /**
     * Delete the stage instance on this channel.
     * @param reason The reason for deleting the stage instance.
     */
    async deleteStageInstance(reason) {
        return this.client.rest.channels.deleteStageInstance(this.id, reason);
    }
    /**
     * Edit the stage instance on this channel.
     * @param options The options for editing the stage instance.
     */
    async editStageInstance(options) {
        return this.client.rest.channels.editStageInstance(this.id, options);
    }
    /**
     * Get the stage instance associated with this channel.
     */
    async getStageInstance() {
        return this.client.rest.channels.getStageInstance(this.id);
    }
    toJSON() {
        return super.toJSON();
    }
}
exports.default = StageChannel;
