// oceanic.js/dist/lib/util/Errors.js
"use strict";
/** @module Errors */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DependencyError = exports.WrapperError = exports.UncachedError = exports.UncaughtError = void 0;
/** An error that is thrown when we encounter an error, and no `error` listeners are present. */
class UncaughtError extends Error {
    name = "UncaughtError";
    constructor(error) {
        super("Uncaught 'error' event", { cause: error });
    }
}
exports.UncaughtError = UncaughtError;
/** An error ancountered when an object is unexpectedly not cached. */
class UncachedError extends Error {
    name = "UncachedError";
    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/ban-types
    constructor(...args) {
        let message;
        if (args.length === 1) {
            [message] = args;
        }
        else {
            const [clazz, property, intent, client] = args;
            message = client && client.options.restMode ? `${clazz.constructor.name}#${property} is not present in rest mode.` : `${clazz.constructor.name}#${property} is not present if you don't have the ${intent} intent.`;
        }
        super(message);
    }
}
exports.UncachedError = UncachedError;
/** An error encountered within {@link InteractionOptionsWrapper | InteractionOptionsWrapper} & {@link SelectMenuValuesWrapper | SelectMenuValuesWrapper}. */
class WrapperError extends Error {
    name = "WrapperError";
    constructor(message) {
        super(message);
    }
}
exports.WrapperError = WrapperError;
/** An error encountered when a needed dependency is missing. */
class DependencyError extends Error {
    name = "DependencyError";
    constructor(message) {
        super(message);
    }
}
exports.DependencyError = DependencyError;
/** A gateway error. */
class GatewayError extends Error {
    code;
    name = "GatewayError";
    constructor(message, code) {
        super(message);
        this.code = code;
    }
}
exports.default = GatewayError;
