// oceanic.js/dist/lib/rest/RESTManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/** @module RESTManager */
const RequestHandler_1 = tslib_1.__importDefault(require("./RequestHandler"));
const Channels_1 = tslib_1.__importDefault(require("../routes/Channels"));
const Guilds_1 = tslib_1.__importDefault(require("../routes/Guilds"));
const Users_1 = tslib_1.__importDefault(require("../routes/Users"));
const OAuth_1 = tslib_1.__importDefault(require("../routes/OAuth"));
const Webhooks_1 = tslib_1.__importDefault(require("../routes/Webhooks"));
const ApplicationCommands_1 = tslib_1.__importDefault(require("../routes/ApplicationCommands"));
const Interactions_1 = tslib_1.__importDefault(require("../routes/Interactions"));
const Routes = tslib_1.__importStar(require("../util/Routes"));
const Miscellaneous_1 = tslib_1.__importDefault(require("../routes/Miscellaneous"));
/** A manager for all rest actions. */
class RESTManager {
    applicationCommands;
    channels;
    #client;
    guilds;
    handler;
    interactions;
    misc;
    oauth;
    users;
    webhooks;
    constructor(client, options) {
        this.applicationCommands = new ApplicationCommands_1.default(this);
        this.channels = new Channels_1.default(this);
        this.#client = client;
        this.guilds = new Guilds_1.default(this);
        this.handler = new RequestHandler_1.default(this, options);
        this.interactions = new Interactions_1.default(this);
        this.misc = new Miscellaneous_1.default(this);
        this.oauth = new OAuth_1.default(this);
        this.users = new Users_1.default(this);
        this.webhooks = new Webhooks_1.default(this);
    }
    get client() {
        return this.#client;
    }
    get options() {
        return this.handler.options;
    }
    /** Alias for {@link RequestHandler#authRequest | RequestHandler#authRequest} */
    async authRequest(options) {
        return this.handler.authRequest(options);
    }
    /**
     * Get the gateway information related to your bot client.
     */
    async getBotGateway() {
        return this.authRequest({
            method: "GET",
            path: Routes.GATEWAY_BOT
        }).then(data => ({
            url: data.url,
            shards: data.shards,
            sessionStartLimit: {
                total: data.session_start_limit.total,
                remaining: data.session_start_limit.remaining,
                resetAfter: data.session_start_limit.reset_after,
                maxConcurrency: data.session_start_limit.max_concurrency
            }
        }));
    }
    /**
     * Get the gateway information.
     */
    async getGateway() {
        return this.request({
            method: "GET",
            path: Routes.GATEWAY
        });
    }
    /** Alias for {@link RequestHandler#request | RequestHandler#request} */
    async request(options) {
        return this.handler.request(options);
    }
}
exports.default = RESTManager;
